lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+    if (status->getSize() != dep.getSize()) return false;
+
+    // Otherwise, if this dependency is verified by modification time, check
+    // it vs. the modification time of the file.
+    if (dep.isModificationTimeBased()) {
+      uint64_t mtime =
+        status->getLastModificationTime().time_since_epoch().count();
+      return mtime == dep.getModificationTime();
+    }
+
+    // Slow path: if the dependency is verified by content hash, check it vs.
+    // the hash of the file.
+    auto buf = getBufferOfDependency(fs, dep.getPath(), interfacePath,
+                                     diags, diagnosticLoc);
+    if (!buf) return false;
+
+    return xxHash64(buf->getBuffer()) == dep.getContentHash();
+  }
+
+  // Check if all the provided file dependencies are up-to-date compared to
+  // what's currently on disk.
+  bool dependenciesAreUpToDate(ArrayRef<FileDependency> deps) {
+    for (auto &in : deps) {
+      if (dependencyTracker)
+        dependencyTracker->addDependency(in.getPath(), /*isSystem*/false);
+      if (!dependencyIsUpToDate(in)) {
+        LLVM_DEBUG(llvm::dbgs() << "Dep " << in.getPath()
+                   << " is directly out of date\n");
+        return false;
+      }
+      LLVM_DEBUG(llvm::dbgs() << "Dep " << in.getPath() << " is up to date\n");
+    }
+    return true;
+  }
+
+  // Check that the output .swiftmodule file is at least as new as all the
+  // dependencies it read when it was built last time.
+  bool serializedASTBufferIsUpToDate(
+    const llvm::MemoryBuffer &buf, SmallVectorImpl<FileDependency> &allDeps) {
+    LLVM_DEBUG(llvm::dbgs() << "Validating deps of " << modulePath << "\n");
+    auto validationInfo = serialization::validateSerializedAST(
+        buf.getBuffer(), /*ExtendedValidationInfo=*/nullptr, &allDeps);
+
+    if (validationInfo.status != serialization::Status::Valid)
+      return false;
+
+    assert(validationInfo.name == moduleName &&
+           "we built a module at this path with a different name?");
+
+    return dependenciesAreUpToDate(allDeps);
+  }
+
+  // Check that the output .swiftmodule file is at least as new as all the
+  // dependencies it read when it was built last time.
+  bool swiftModuleIsUpToDate(StringRef modulePath,
+                             SmallVectorImpl<FileDependency> &AllDeps) {
+    auto OutBuf = fs.getBufferForFile(modulePath);
+    if (!OutBuf)
+      return false;
+    return serializedASTBufferIsUpToDate(*OutBuf.get(), AllDeps);
+  }
+
+  // Check that a "forwarding" .swiftmodule file is at least as new as all the
+  // dependencies it read when it was built last time. Requires that the
+  // forwarding module has been loaded from disk.
+  bool forwardingModuleIsUpToDate(const ForwardingModule &fwd,
+                                  SmallVectorImpl<FileDependency> &deps) {
+    // First, make sure the underlying module path exists and is valid.
+    auto modBuf = getSerializedASTBuffer(fs, fwd.underlyingModulePath);
```

Comment:
If we're importing the module, this causes us to read the contents twice, no? Seems like a TOCTOU error.

Reply:
I left a FIXME in place for the real fix here, which is to ensure we only open a buffer once. I will address that in a follow-up.

Reply:
rdar://48654608